nabs/utils.py

```diff
@@ -110,3 +116,119 @@ def add_named_kwargs_to_signature(
     )
 
     return sig.replace(parameters=start_params + wrapper_params + end_params)
+
+
+def format_ophyds_to_html(obj, allow_child=False):
+    """
+    Recursively construct html that contains the output from .status() for
+    each object provided.  Base case is being passed a single ophyd object
+    with a `.status()` method.  Any object without a `.status()` method is
+    ignored.
+
+    Creates divs and buttons based on styling
+    from `nabs._html.collapse_list_head` and `nabs._html.collapse_list_tail`
+
+    Parameters
+    ----------
+    obj : ophyd object or Iterable of ophyd objects
+        Objects to format into html
+
+    allow_child : bool, optional
+        Whether or not to post child devices to the elog.  Defaults to False,
+        to keep long lists of devices concise
+
+    Returns
+    -------
+    out : string
+        html body containing ophyd object representations (sans styling, JS)
+    """
+    if isinstance(obj, Iterable):
+        content = ""
+
+        for o in obj:
+            content += format_ophyds_to_html(o, allow_child=allow_child)
+
+        # Don't return wrapping if there's no content
+        if content == "":
+            return content
+
+        # HelpfulNamespaces tend to lack names, maybe they won't some day
+        parent_default = ('Ophyd status: ' +
+                          ', '.join('[...]' if isinstance(o, Iterable)
+                                    else o.name for o in obj))
+        parent_name = getattr(obj, '__name__', parent_default[:60] + ' ...')
+
+        # Wrap in a parent div
+        out = (
+            "<button class='collapsible'>" +
+            f"{parent_name}" +  # should be a namespace name
+            "</button><div class='parent'>" +
+            content +
+            "</div>"
+        )
+        return out
+
+    # check if parent level ophyd object
+    elif (callable(getattr(obj, 'status', None)) and
+            ((getattr(obj, 'parent', None) is None and
+              getattr(obj, 'biological_parent', None) is None) or
+             allow_child)):
+        content = ""
+        try:
+            content = (
+                f"<button class='collapsible'>{obj.name}</button>" +
+                f"<div class='child content'><pre>{obj.status()}</pre></div>"
+            )
+        except Exception as ex:
+            logger.info(f'skipped {str(obj)}, due to Exception: {ex}')
+
+        return content
+
+    # fallback base case (if ignoring obj)
+    else:
+        return ""
+
+
+def post_ophyds_to_elog(elog, objs, allow_child=False):
```

Comment:
Minor nitpick: no type annotations anywhere, though I don't think these are required.

Reply:
We are generally moving toward annotating most new code, @tangkong, just so you know.
It'll take some getting used to, but we're happy to help when you need it.